Allow rendering other modules from placeholders (#3900)
GregTechCEu/GregTechCEu-1.19: docs/content/Modpacks/Other-Topics/Central-Monitor.md

---
title: Central Monitor & Placeholder System
---

!!! info "But can it run Doom?"

>     **YES.** <h6>_If anyone is crazy enough to program Doom in the placeholder language._</h6>

Reviewer: remove this one as well

### Custom monitor modules
If you want to add a monitor module, simply attach a component that implements `IMonitorModuleItem` to your `ComponentItem`.
Modules can have a custom UI, can be ticked (in a placeholder or not) and, most importantly, rendered.
??? example "Example of a custom module in Java"
    ```java
    public class ExampleModuleBehaviour implements IMonitorModuleItem {
        @Override
        public String getType() {
            // can be any string, this is currently only used for CC: Tweaked compat
            return "example";
        }

        @Override
        public void tick(ItemStack stack, CentralMonitorMachine machine, MonitorGroup group) {

>             // this is only run on the logical server side

Reviewer: ```suggestion
            // this is only called on the logical server
```

            // put all of your module's logic here instead of in getRenderer(stack)
            // can also be left completely empty (like in the image module)
        }

        @Override
        public void tickInPlaceholder(ItemStack stack, PlaceholderContext context) {
            // this is also only run on the logical server side
            // but this is only called when a placeholder accesses this module (and wants to render it)
            // this is not called on each tick

>             // you can even put the same code here as in the tick() method (like the text module)

Reviewer: ```suggestion
            // this is also only called on the logical server, but only when a placeholder accesses this module and wants to render it
            // this *isn't* called on each tick
            // you can even put the same code here as in the tick() method, like the text module does
```

        }

        @Override
        public IMonitorRenderer getRenderer(ItemStack stack) {
            // this is only called on the logical client side
            // should return a new instance of the renderer for this module (not null)

>             // for examples of renderer code look in the GTCEu Modern github (src/main/java/com/gregtechceu/gtceu/client/renderer)

Reviewer: ```suggestion
            // this is only called on the logical client
            // should return a new instance of the renderer for this module (not null)
            // for examples of renderer code look in the GTCEu Modern github:
            // https://github.com/GregTechCEu/GregTech-Modern/tree/1.20.1/src/main/java/com/gregtechceu/gtceu/client/renderer
```

All placeholders work on strings (or, more specifically, `Component`s to allow text formatting), so when you write `{calc {calc 2 + 4} * 3}`,
first `{calc 2 + 4}` will be evaluated into `6`, then it will be converted to a string and back to an int, and then it will be passed into the second placeholder
to evaluate `{calc 6 * 3}` into `18`, which will be turned into a string again. That also allows for things like `{calc 3 + 1}2`, which will evaluate into `42`,
since outside of placeholders text is simply concatenated. Placeholder arguments are separated by spaces, which may be a bit annoying, when wanting to pass a string
with a space into a placeholder, for example `{if 1 string with spaces}`, which will cause an error. In these cases, double quotes can be used: `{if 1 "string with spaces"}`
will work perfectly fine. There are placeholders that need reference items, to achieve that, there are 8 slots in the text module's UI on the left.
Items can be inserted/extracted from these slots automatically using the `ender` placeholder by interacting with Ender Item Links.<br>

> !!! tip "The full list of placeholders with explanations on what they do and usage examples can be found in-game in the text module or computer monitor UI on the left."

Reviewer 2: Maybe replace this with something like "To read more about placeholders, read *here*[link to other page]" so we don't have text duplication
